[macOS Sonoma] Add a sparse SDK for public SDK overrides

emw-apple opened the pull request:
#### 6dba0b8967fcd4a9c8c28c242395eda798fe6caf
<pre>
[macOS Sonoma] Add a sparse SDK for public SDK overrides
<a href="https://bugs.webkit.org/show_bug.cgi?id=258749">https://bugs.webkit.org/show_bug.cgi?id=258749</a>
rdar://problem/111922548

Reviewed by Alexey Proskuryakov.

WebKit uses SPI-only classes in the SDK which are marked
API_UNAVAILABLE. Instead of handling this in
configure-xcode-for-embedded-development, add a sparse SDK which
contains an AvailabilityProhibitedInternal.h header that disables
unavailability annotations.

This establishes precedent for making more SDK-level customizations via
sparse SDKs.

* Configurations/CommonBase.xcconfig:
* Configurations/SDKAdditions.xcconfig: Added, contains conditional
  logic to determine when a sparse SDK should be used. Defines
  WK_ADDITIONAL_SDKS, which is used in projects to set ADDITIONAL_SDKS.
* Configurations/WebKitProjectPaths.xcconfig: Refactor the logic used to
  figure out the path to WebKitLibraries/. Previously, it assumed any
  project was two directories deep in the repo hierarchy (i.e.
  Source/WebCore/WebCore.xcodeproj). Now, it figures out the workspace
  root path dynamically based on SRCROOT.

In projects, set ADDITIONAL_SDKS = $(WK_ADDITIONAL_SDKS):

* PerformanceTests/DecoderTest/Configurations/Base.xcconfig:
* PerformanceTests/MediaTime/Configurations/Base.xcconfig:
* Source/JavaScriptCore/Configurations/Base.xcconfig:
* Source/ThirdParty/ANGLE/Configurations/Base.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/Base.xcconfig:
* Source/WTF/Configurations/Base.xcconfig:
* Source/WebCore/Configurations/Base.xcconfig:
* Source/WebCore/PAL/Configurations/Base.xcconfig:
* Source/WebCore/PAL/ThirdParty/libavif/Configurations/Base.xcconfig:
* Source/WebCore/PAL/ThirdParty/libavif/ThirdParty/dav1d/Configurations/Base.xcconfig:
* Source/WebGPU/Configurations/Base.xcconfig:
* Source/WebInspectorUI/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/SwiftOverlay/Configurations/Base.xcconfig:
* Source/WebKitLegacy/mac/Configurations/Base.xcconfig:
* Source/bmalloc/Configurations/Base.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/Base.xcconfig:
* Tools/ImageDiff/cg/Configurations/Base.xcconfig:
* Tools/MiniBrowser/Configurations/Base.xcconfig:
* Tools/MiniBrowserSwiftUI/Configurations/Base.xcconfig:
* Tools/MobileMiniBrowser/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/WebEditingTester/Configurations/Base.xcconfig:
* Tools/WebGPUPlayground/Configurations/Base.xcconfig:
* Tools/WebKitTestRunner/Configurations/Base.xcconfig:
* Tools/lldb/lldbWebKitTester/Configurations/Base.xcconfig:

The sparse SDK contains an AvailabilityProhibitedInternal header,
exactly the same as the one installed by
configure-xcode-for-embedded-development.

* WebKitLibraries/SDKs/macosx14.0-additions.sdk/SDKSettings.plist: Added.
* WebKitLibraries/SDKs/macosx14.0-additions.sdk/usr/local/include/AvailabilityProhibitedInternal.h: Added.

Canonical link: <a href="https://commits.webkit.org/265967@main">https://commits.webkit.org/265967@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/20c30152cdb12925c0bf0269ffdcdfd03f123317

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12400 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12735 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13046 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14142 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11925 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12459 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15161 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12753 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14594 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12564 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13322 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10489 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14565 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10605 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11234 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18329 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/10552 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11688 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11399 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14582 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/11738 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11889 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/9816 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/12461 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11119 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3267 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3056 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15450 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/12815 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11744 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3047 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Configurations/WebKitProjectPaths.xcconfig:
```diff
@@ -73,14 +73,25 @@ UMBRELLA_FRAMEWORKS_DIR_OVERRIDE_YES = $(WK_OVERRIDE_FRAMEWORKS_DIR);
 WK_PRIVATE_FRAMEWORK_STUBS_DIR = $(WK_PRIVATE_FRAMEWORK_STUBS_DIR_$(USE_INTERNAL_SDK));
-WK_PRIVATE_FRAMEWORK_STUBS_DIR_[sdk=macos*] = $(PROJECT_DIR)/../../WebKitLibraries/WebKitPrivateFrameworkStubs/Mac/$(TARGET_MAC_OS_X_VERSION_MAJOR);
-WK_PRIVATE_FRAMEWORK_STUBS_DIR_[sdk=iphone*] = $(PROJECT_DIR)/../../WebKitLibraries/WebKitPrivateFrameworkStubs/iOS/$(WK_TARGET_IOS_VERSION_MAJOR);
-WK_PRIVATE_FRAMEWORK_STUBS_DIR_[sdk=appletv*] = $(PROJECT_DIR)/../../WebKitLibraries/WebKitPrivateFrameworkStubs/appletvos/$(WK_TARGET_TVOS_VERSION_MAJOR);
-WK_PRIVATE_FRAMEWORK_STUBS_DIR_[sdk=watch*] = $(PROJECT_DIR)/../../WebKitLibraries/WebKitPrivateFrameworkStubs/watchos/$(WK_TARGET_WATCHOS_VERSION_MAJOR);
+WK_PRIVATE_FRAMEWORK_STUBS_DIR_[sdk=macos*] = $(WK_WORKSPACE_DIR)/WebKitLibraries/WebKitPrivateFrameworkStubs/Mac/$(TARGET_MAC_OS_X_VERSION_MAJOR);
+WK_PRIVATE_FRAMEWORK_STUBS_DIR_[sdk=iphone*] = $(WK_WORKSPACE_DIR)/WebKitLibraries/WebKitPrivateFrameworkStubs/iOS/$(WK_TARGET_IOS_VERSION_MAJOR);
+WK_PRIVATE_FRAMEWORK_STUBS_DIR_[sdk=appletv*] = $(WK_WORKSPACE_DIR)/WebKitLibraries/WebKitPrivateFrameworkStubs/appletvos/$(WK_TARGET_TVOS_VERSION_MAJOR);
+WK_PRIVATE_FRAMEWORK_STUBS_DIR_[sdk=watch*] = $(WK_WORKSPACE_DIR)/WebKitLibraries/WebKitPrivateFrameworkStubs/watchos/$(WK_TARGET_WATCHOS_VERSION_MAJOR);
 
 WK_DOWNLEVEL_FRAMEWORK_STUBS_DIR[sdk=macos*] = $(WK_DOWNLEVEL_FRAMEWORK_STUBS_DIR$(WK_MACOS_1400);
-WK_DOWNLEVEL_FRAMEWORK_STUBS_DIR_MACOS_BEFORE_1400 = $(PROJECT_DIR)/../../WebKitLibraries/DownlevelFrameworkStubs/$(TARGET_MAC_OS_X_VERSION_MAJOR);
+WK_DOWNLEVEL_FRAMEWORK_STUBS_DIR_MACOS_BEFORE_1400 = $(WK_WORKSPACE_DIR)/WebKitLibraries/DownlevelFrameworkStubs/$(TARGET_MAC_OS_X_VERSION_MAJOR);
```
This is a small bug fix: PAL and other projects that were not top-level projects in the `Source` directory had an incorrect path to framework stubs.

aproskuryakov commented on Configurations/WebKitProjectPaths.xcconfig:
```diff
@@ -73,14 +73,25 @@ UMBRELLA_FRAMEWORKS_DIR_OVERRIDE_YES = $(WK_OVERRIDE_FRAMEWORKS_DIR);
-WK_DOWNLEVEL_FRAMEWORK_STUBS_DIR_MACOS_BEFORE_1400 = $(PROJECT_DIR)/../../WebKitLibraries/DownlevelFrameworkStubs/$(TARGET_MAC_OS_X_VERSION_MAJOR);
+WK_DOWNLEVEL_FRAMEWORK_STUBS_DIR_MACOS_BEFORE_1400 = $(WK_WORKSPACE_DIR)/WebKitLibraries/DownlevelFrameworkStubs/$(TARGET_MAC_OS_X_VERSION_MAJOR);
+
+WK_PUBLIC_SDK_ADDITIONS_PATH = $(WK_WORKSPACE_DIR)/WebKitLibraries/SDKs/WebKitAdditions.$(PLATFORM_NAME)$($(DEPLOYMENT_TARGET_SETTING_NAME):base).sdk;
```
Should this be `WebKitSDKAdditions<...>.sdk`, so that there isn't a name conflict with the very unrelated WebKitAdditions project?

emw-apple replied:
Maybe a better naming convension is `$(PLATFORM_NAME)$($(DEPLOYMENT_TARGET_SETTING_NAME):base)Additions`, to get directories like "macosx14Additions.sdk". (The case inconsistency between SDK name and platform name is annoying, but there's not much we can do.)

aproskuryakov commented on Source/JavaScriptCore/Configurations/DebugRelease.xcconfig:
```diff
@@ -37,6 +37,8 @@ CODE_SIGN_IDENTITY_YES = $(WK_ENGINEERING_CODE_SIGN_IDENTITY);
 SDKROOT = $(SDKROOT_$(USE_INTERNAL_SDK));
 SDKROOT_ = macosx;
 SDKROOT_YES = macosx.internal;
+ADDITIONAL_SDKS = $(ADDITIONAL_SDKS_$(USE_INTERNAL_SDK));
```
Any chance that we'll ever want `$(inherit)` here? Probably best to add it everywhere preemptively if so.

emw-apple replied:
Good idea. I'd like to extend this mechanism for downlevels, at which point we'd need a declaration in Base.xcconfig.

aproskuryakov reviewed (approved):
Nice. I understand that this is laying down the basic framework, but as it stands, this could apply to all platforms equally, so we don't need a version name to keep updating, and can remove this part from `configure-xcode-for-embedded-development`. Not sure about the tradeoff to make sparse SDKs not just per platform, but also per SDK version. Unless you plan to move all `.tbds` into sparse SDKs?

aproskuryakov commented on PerformanceTests/DecoderTest/Configurations/Base.xcconfig:
```diff
@@ -120,6 +120,7 @@ DEAD_CODE_STRIPPING_normal = YES;
 DEAD_CODE_STRIPPING = $(DEAD_CODE_STRIPPING_$(CURRENT_VARIANT));
 
 SDKROOT = macosx.internal;
+ADDITIONAL_SDKS = $(WK_ADDITIONAL_SDKS);
```
What is the reason to indirectly define `ADDITIONAL_SDKS` via `WK_ADDITIONAL_SDKS`? Can we define `ADDITIONAL_SDKS` in `SDKAdditions.xcconfig`?

Also, while this avoids the need to inherit from Base to DebugRelease that was discussed earlier, are there any other sources to inherit from, like command line perhaps?

emw-apple committed:
[macOS Sonoma] Add a sparse SDK for public SDK overrides
https://bugs.webkit.org/show_bug.cgi?id=258749
rdar://problem/111922548

Reviewed by Alexey Proskuryakov.

WebKit uses SPI-only classes in the SDK which are marked
API_UNAVAILABLE. Instead of handling this in
configure-xcode-for-embedded-development, add a sparse SDK which
contains an AvailabilityProhibitedInternal.h header that disables
unavailability annotations.

This establishes precedent for making more SDK-level customizations via
sparse SDKs.

* Configurations/CommonBase.xcconfig:
* Configurations/SDKAdditions.xcconfig: Added, contains conditional
  logic to determine when a sparse SDK should be used. Defines
  WK_ADDITIONAL_SDKS, which is used in projects to set ADDITIONAL_SDKS.
* Configurations/WebKitProjectPaths.xcconfig: Refactor the logic used to
  figure out the path to WebKitLibraries/. Previously, it assumed any
  project was two directories deep in the repo hierarchy (i.e.
  Source/WebCore/WebCore.xcodeproj). Now, it figures out the workspace
  root path dynamically based on SRCROOT.

In projects, set ADDITIONAL_SDKS = $(WK_ADDITIONAL_SDKS):

* PerformanceTests/DecoderTest/Configurations/Base.xcconfig:
* PerformanceTests/MediaTime/Configurations/Base.xcconfig:
* Source/JavaScriptCore/Configurations/Base.xcconfig:
* Source/ThirdParty/ANGLE/Configurations/Base.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/Base.xcconfig:
* Source/WTF/Configurations/Base.xcconfig:
* Source/WebCore/Configurations/Base.xcconfig:
* Source/WebCore/PAL/Configurations/Base.xcconfig:
* Source/WebCore/PAL/ThirdParty/libavif/Configurations/Base.xcconfig:
* Source/WebCore/PAL/ThirdParty/libavif/ThirdParty/dav1d/Configurations/Base.xcconfig:
* Source/WebGPU/Configurations/Base.xcconfig:
* Source/WebInspectorUI/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/SwiftOverlay/Configurations/Base.xcconfig:
* Source/WebKitLegacy/mac/Configurations/Base.xcconfig:
* Source/bmalloc/Configurations/Base.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/Base.xcconfig:
* Tools/ImageDiff/cg/Configurations/Base.xcconfig:
* Tools/MiniBrowser/Configurations/Base.xcconfig:
* Tools/MiniBrowserSwiftUI/Configurations/Base.xcconfig:
* Tools/MobileMiniBrowser/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/WebEditingTester/Configurations/Base.xcconfig:
* Tools/WebGPUPlayground/Configurations/Base.xcconfig:
* Tools/WebKitTestRunner/Configurations/Base.xcconfig:
* Tools/lldb/lldbWebKitTester/Configurations/Base.xcconfig:

The sparse SDK contains an AvailabilityProhibitedInternal header,
exactly the same as the one installed by
configure-xcode-for-embedded-development.

* WebKitLibraries/SDKs/macosx14.0-additions.sdk/SDKSettings.plist: Added.
* WebKitLibraries/SDKs/macosx14.0-additions.sdk/usr/local/include/AvailabilityProhibitedInternal.h: Added.

Canonical link: https://commits.webkit.org/265967@main
